#Recomendación

## Importación libreria

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

import os

import warnings
warnings.filterwarnings("ignore")

# time calculation to track some processes
import time

import pickle

# import the function to compute cosine_similarity
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.cluster import KMeans

import scipy as sp

## Importación datos

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
ruta= '/content/drive/MyDrive/Colab Notebooks/Nuclio - Data Science/TFM/df_grande_clusters.pkl'
ruta2= '/content/drive/MyDrive/Colab Notebooks/Nuclio - Data Science/TFM/Pickle_clusters.pkl'
PATH_DATA = '/content/drive/MyDrive/Colab Notebooks/Nuclio - Data Science/TFM/'

Pickle del clustering - con todas las columnas

In [4]:
import pickle

with open(ruta, 'rb') as handle:
    df_info = pickle.load(handle)

Pickle del clustering - solo con las columnas del cluster

In [5]:
import pickle

with open(ruta2, 'rb') as handle:
    df_group = pickle.load(handle)

Datos de los productos

In [6]:
products = pd.read_csv(PATH_DATA+'products_df.csv', encoding='utf-8')
products.drop(columns=['Unnamed: 0'], inplace=True)

In [7]:
df_group.head()

age  meses_recency  ...  financiacion_uso_medio  cluster
pk_cid                      ...                                 
15891    59             -1  ...                       0        2
16063    62             -1  ...                       0        2
16203    70              4  ...                       0        2
16502    58              4  ...                       0        2
17457    54              0  ...                       0        5

[5 rows x 12 columns]

In [8]:
df_info.head()

pk_cid  active_mean  ... ahorro_uso_medio  financiacion_uso_medio
0   15891     0.500000  ...         0.000000                0.000000
1   16063     0.714286  ...         0.000000                0.000000
2   16203     0.833333  ...         0.000000                0.000000
3   16502     0.888889  ...         0.000000                0.000000
4   17457     1.000000  ...         0.941176                0.294118

[5 rows x 71 columns]

In [9]:
products.head()

pk_cid pk_partition  ...  em_account_p  em_acount
0  1375586   2018-01-28  ...             0          1
1  1050611   2018-01-28  ...             0          1
2  1050612   2018-01-28  ...             0          1
3  1050613   2018-01-28  ...             0          0
4  1050614   2018-01-28  ...             0          1

[5 rows x 17 columns]

## Construcción dataset recomendación

Unimos productos con clusters

In [10]:
df_group = df_group.reset_index()

In [11]:
full_df = pd.merge(products,df_group[['pk_cid','cluster']],on='pk_cid',how='left')

Miramos solo los clusters 4 y 5, ya que son los que más gastan y estos serán nuestros targets

In [12]:
full_df = full_df[(full_df['cluster']==4) | (full_df['cluster']==5)]

Calculamos los productos que tiene cada cliente

In [13]:
full_df["total"] = full_df.iloc[:,2:-1].sum(axis=1)

Calculamos el total de todas las altas que han habido

In [14]:
test = full_df.groupby('pk_cid')['total'].sum()

In [15]:
full_df = pd.merge(full_df,test,on = 'pk_cid')

In [16]:
full_df = full_df.rename(columns={'total_x': 'total_mes', 'total_y': 'total_sum'})

Eliminamos la columna payroll ya que no aporta beneficios 


In [17]:
full_df = full_df[['pk_cid','pk_partition','short_term_deposit', 'loans', 'mortgage',
       'funds', 'securities', 'long_term_deposit',
       'credit_card', 'pension_plan', 'payroll_account',
       'emc_account', 'debit_card', 'em_account_p', 'em_acount','cluster','total_mes','total_sum']]

Agrupamos por la suma de todos los productos que ha contratado cada cliente en EasyMoney y ordenamos por orden descendente para quedarnos con los clientes que más productos han contratado con EasyMoney

In [18]:
cliente_total = full_df.groupby('pk_cid')[['total_sum']].max()

In [19]:
cliente_total = cliente_total.sort_values('total_sum',ascending = False)

In [20]:
cliente_total

total_sum
pk_cid            
1128353      149.0
1190607      145.0
1070525      142.0
1233988      136.0
1054409      136.0
...            ...
1255271        0.0
1020270        0.0
1020271        0.0
1390887        0.0
1553552        0.0

[65234 rows x 1 columns]

Tenemos que enviar 10.000 emails -> nos quedamos con los 10.000 clientes que más han contratado

In [21]:
cliente_mail = cliente_total[:10100] #filtramos por 100 clientes más porque hay clientes que no estan en el ultimo pk_partition

Filtramos la lista de clientes en el Dataframe principal

In [22]:
cliente_mail

total_sum
pk_cid            
1128353      149.0
1190607      145.0
1070525      142.0
1233988      136.0
1054409      136.0
...            ...
1059321       49.0
1224656       49.0
1251684       49.0
1084701       49.0
1163267       49.0

[10100 rows x 1 columns]

In [23]:
target_clients = full_df[full_df['pk_cid'].isin(cliente_mail.index)]

In [24]:
target_clients['pk_cid'].nunique()

10100

Miramos que productos ha contratado cada cliente durante el tiempo que ha estado con nosotros

In [25]:
target_df = target_clients.groupby('pk_cid')[['short_term_deposit', 'loans', 'mortgage',
       'funds', 'securities', 'long_term_deposit',
       'credit_card', 'pension_plan', 'payroll_account',
       'emc_account', 'debit_card', 'em_account_p', 'em_acount']].sum()

2 approaches: 
  

*   Mirar que productos suele contratar el cliente y mirar si actualmente tiene contratado esos productos -> si no lo tiene, recomendarle el producto que más haya contratado con nosotros
*   Mirar que productos contratan esos clientes y hacer un ranking de los productos e ir recomendando los productos más demandado que no tenga contratado un cliente.



Nos decantamos por el 2n approach

Reducimos el tamaño

In [26]:
target_df = target_df.astype('int8')

Cosine similarity - obtenemos matriz de cosinus para product based y client based similarity

In [27]:
FILE_NO_EXISTS = False
if FILE_NO_EXISTS:

    st = time.time()
    target_df_norm = target_df.apply(lambda x: (x - np.mean(x))/(np.max(x) - np.min(x)), axis = 1)
    et = time.time()
    print("Normalization took {} minutes".format(round((et - st)/60, 2)))

    sparse_ratings = sp.sparse.csr_matrix(target_df_norm.values)

    st = time.time()
    
    product_similarity = cosine_similarity(sparse_ratings.T)
    product_sim_df = pd.DataFrame(product_similarity, index = target_df_norm.columns, columns = target_df_norm.columns)
    product_sim_df.to_pickle(os.path.join(PATH_DATA, "product_similarity.pkl"))

    user_similarity = cosine_similarity(sparse_ratings)
    user_sim_df = pd.DataFrame(user_similarity, index = target_df_norm.index, columns = target_df_norm.index)
    user_sim_df.to_pickle(os.path.join(PATH_DATA, "user_similarity.pkl"))
    
    et = time.time()
    print("Total time to calculate similarity took {} minutes.".format(round((et - st)/60, 2)))
    
else:
    user_sim_df = pd.read_pickle(os.path.join(PATH_DATA, "user_similarity.pkl"))
    product_sim_df = pd.read_pickle(os.path.join(PATH_DATA, "product_similarity.pkl"))

In [28]:
user_sim_df.head()

pk_cid   17970     26018     31624    ...   1437442   1442421   1479480
pk_cid                                ...                              
17970   1.000000 -0.404410  0.813570  ...  0.357907  0.491531  0.652432
26018  -0.404410  1.000000 -0.321171  ...  0.231025 -0.046819  0.227184
31624   0.813570 -0.321171  1.000000  ...  0.529662  0.706904  0.521436
32376   0.809146 -0.299820  0.998417  ...  0.554826  0.730903  0.514566
45370   0.603646  0.027397  0.402521  ...  0.344641  0.533826  0.382195

[5 rows x 10100 columns]

In [29]:
product_sim_df

short_term_deposit     loans  ...  em_account_p  em_acount
short_term_deposit            1.000000  0.985296  ...      0.993868   0.021773
loans                         0.985296  1.000000  ...      0.990058   0.018157
mortgage                      0.987423  0.983642  ...      0.992167   0.016371
funds                         0.749588  0.743112  ...      0.750767   0.145353
securities                    0.739721  0.733349  ...      0.741259   0.108068
long_term_deposit             0.505670  0.494539  ...      0.500406   0.239637
credit_card                  -0.013758 -0.004097  ...     -0.010316  -0.081653
pension_plan                 -0.721439 -0.717800  ...     -0.721448  -0.448980
payroll_account              -0.791690 -0.787756  ...     -0.791910  -0.455206
emc_account                  -0.457159 -0.458706  ...     -0.459734   0.188062
debit_card                   -0.839971 -0.834433  ...     -0.840602  -0.114588
em_account_p                  0.993868  0.990058  ...      1.000000   0.018890
em_acount                     0.021773  0.018157  ...      0.018890   1.000000

[13 rows x 13 columns]

Dado que hay muchos clientes similares por tener los mismos productos contratado, se opta por utilizar la matriz según la similitud de productos

##Recomendación product based

Miramos que tiene cada cliente contratado actualmente

In [30]:
target_clients_now = target_clients[target_clients['pk_partition']==target_clients['pk_partition'].max()]

In [31]:
target_clients_now

pk_cid pk_partition  short_term_deposit  ...  cluster  total_mes  total_sum
118     1050686   2019-05-28                   0  ...        5        3.0       50.0
135     1050660   2019-05-28                   0  ...        5        5.0       76.0
169     1050651   2019-05-28                   0  ...        5        5.0       75.0
186     1050652   2019-05-28                   0  ...        5        4.0       78.0
203     1050659   2019-05-28                   0  ...        5        7.0       93.0
...         ...          ...                 ...  ...      ...        ...        ...
866558  1433369   2019-05-28                   0  ...        4        6.0       56.0
867166  1428596   2019-05-28                   0  ...        4        6.0       51.0
869437  1437442   2019-05-28                   0  ...        4        7.0       59.0
883713  1442421   2019-05-28                   0  ...        4        6.0       50.0
895625  1479480   2019-05-28                   0  ...        4        7.0       55.0

[10072 rows x 18 columns]

Función para sacar los top productos que esta relacionado con otro producto

Esta función se utiliza más adelante y servirá para saber que 4 productos está más relacionado a un producto en concreto

In [32]:
def top_products(name, df):
    '''
    This functions prints top 10 similar products, based on the number of times that they were acquired.
    '''
    #print('Similar shows to {} include:\n'.format(name))
    
    index = product_sim_df[name].sort_values(ascending = False).index[1:11]
    values = product_sim_df[name].sort_values(ascending = False).values[1:11]

    return {index[0]:values[0],index[1]:values[1],index[2]:values[2],index[3]:values[3]}


Hacemos drop de las columnas que no necesitamos para la recomendación

In [33]:
targets_clients_now_df = target_clients_now.drop(['pk_partition','total_mes','total_sum','cluster'],axis=1)

In [34]:
target_clients_now[target_clients_now['total_mes']==0]

pk_cid pk_partition  short_term_deposit  ...  cluster  total_mes  total_sum
20958   1045535   2019-05-28                   0  ...        4        0.0       88.0
73030   1020461   2019-05-28                   0  ...        4        0.0       68.0
74847   1020757   2019-05-28                   0  ...        4        0.0       62.0
114193  1123356   2019-05-28                   0  ...        4        0.0       55.0
128972  1122099   2019-05-28                   0  ...        4        0.0       62.0
169503  1119050   2019-05-28                   0  ...        4        0.0       75.0
176542  1116106   2019-05-28                   0  ...        4        0.0       62.0
418412  1378440   2019-05-28                   0  ...        4        0.0       55.0
565434  1139241   2019-05-28                   0  ...        4        0.0       60.0
610001  1223301   2019-05-28                   0  ...        4        0.0       51.0
653019  1236017   2019-05-28                   0  ...        4        0.0       60.0
693930  1188295   2019-05-28                   0  ...        4        0.0       54.0
700517  1209899   2019-05-28                   0  ...        4        0.0       53.0
730373   471701   2019-05-28                   0  ...        4        0.0       50.0

[14 rows x 18 columns]

In [35]:
targets_clients_now_df = targets_clients_now_df.set_index('pk_cid')

Construimos un diccionario de que productos se recomendaría a cada cliente


*   Miramos que productos tiene contratado cada cliente a día de hoy (última fecha disponible)
*   Para cada uno de esos productos, nos sacamos los 4 productos que más se parece, que serán los que vamos a recomendar después



In [36]:
from collections import OrderedDict
def get_dic(df):
  dict_client = {}
  list_dict = []
  for cliente in df.index:
    list_client = []
    list_dict = []
    dict_product = {}
    for product,value in zip((df[df.index == cliente].any() == True).index,(df[df.index == cliente].any() == True).values) :
      if value == True:
        dict_top = top_products(product,product_sim_df)
        for element in dict_top:
          if element in dict_product:
            if dict_top[element] > dict_product[element]:
              dict_product[element] = dict_top[element]
          else:
            dict_product[element] = dict_top[element]
    sorted_d = OrderedDict( sorted(dict_product.items(),
                           key=lambda item: item[1],
                           reverse=True))
    dict_client[cliente] = sorted_d
  return dict_client

El diccionario de cada cliente y que productos le vamos a recomendar de mayor similitud a menor según los productos que tiene ese cliente contratado

In [37]:
diccionario_cliente = get_dic(targets_clients_now_df)

Construimos el dataframe con el cliente y que producto que vamos a recomendar

In [38]:
lista_recomendacion = pd.DataFrame(columns=['Cliente','Recomendacion'])
for cliente in targets_clients_now_df.index:
  for producto in diccionario_cliente[cliente]:
      if (targets_clients_now_df[targets_clients_now_df.index == cliente][producto].to_list())[0] == 0:
        lista_recomendacion = lista_recomendacion.append({'Cliente':cliente,'Recomendacion':producto},ignore_index = True)
        break
 

In [39]:
lista_recomendacion['Recomendacion'].value_counts()

emc_account           2993
pension_plan          1870
em_acount             1664
payroll_account       1298
debit_card             911
long_term_deposit      524
em_account_p           501
funds                  290
loans                    6
short_term_deposit       1
Name: Recomendacion, dtype: int64

Salen menos clientes porque hay clientes que no tienen ningún producto contratado y no se le puede asignar una recomendación 

In [40]:
lista_recomendacion

Cliente      Recomendacion
0      1050686    payroll_account
1      1050660          em_acount
2      1050651          em_acount
3      1050652       pension_plan
4      1050659       em_account_p
...        ...                ...
10053  1433369          em_acount
10054  1428596          em_acount
10055  1437442       em_account_p
10056  1442421  long_term_deposit
10057  1479480       em_account_p

[10058 rows x 2 columns]

Como hay clientes que no existen en el último pk_partition y hay otros que no tienen ningún producto contratado. Partiamos de 10100 clientes y el resultado es 10058. Haremos un drop de 58 clientes con em_acount_p que es el que nos da menos beneficio

In [41]:
len_lista=  len(lista_recomendacion)
for cliente in lista_recomendacion.Cliente:
  if len_lista != 10000:
    if (lista_recomendacion[lista_recomendacion.Cliente == cliente].Recomendacion == 'em_account_p').to_list()[0]:
      lista_recomendacion.drop(lista_recomendacion.loc[lista_recomendacion['Cliente']==cliente].index, inplace=True)
      len_lista = len_lista -1

In [42]:
lista_recomendacion['Recomendacion'].value_counts()

emc_account           2993
pension_plan          1870
em_acount             1664
payroll_account       1298
debit_card             911
long_term_deposit      524
em_account_p           443
funds                  290
loans                    6
short_term_deposit       1
Name: Recomendacion, dtype: int64